### PR TITLE
add mod extension for editing mod.json

### PR DESCRIPTION
**Pull request by NikkyAI**
<s>this currently only works on a existing `mod.json`

properties that can be set: `id`, `version`

planned: everything else

properties currently just override what happened to be in the mod.json, this could be changed to allow appending or prepending values
I think this would be overengineered
such things can also just stay in the `build.gradle`
</s>

new approach: dumb token replacement

available methods on the mod extension:
- `replace(String key, String value)`
- `replace(Map<String, String> replacements)`

sample usage

```kotlin
mod {
   replace("VERSION", version + "-" + buildNumber)
}
```

`mod.json`
```json
{
  "id": "fabric-language-kotlin",
  "version": "VERSION"
}
```

### CODE REVIEW

File: src/main/java/net/fabricmc/loom/ModExtension.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package net.fabricmc.loom;
+
+import org.gradle.api.Project;
+
+public class ModExtension {
+    private final Project project;
```
Comment (jamierocks) [inline src/main/java/net/fabricmc/loom/ModExtension.java]: Just need to fix the formatting on this file (tabs).

**Review by jamierocks — CHANGES_REQUESTED** [review]
Looking good :+1:

File: src/main/java/net/fabricmc/loom/ModExtension.java (hunk 1 above)
Comment (NikkyAI) [reply src/main/java/net/fabricmc/loom/ModExtension.java]: intereesting.. tabs and spaces seem to be used interchangeably in the project
net.fabricmc.loom.util.MixinRefmapHelper for example is spaces ?

Comment (NikkyAI) [reply src/main/java/net/fabricmc/loom/ModExtension.java]: i fixed that and all other occurences of 4 spaces that should be tabs

**Comment by NikkyAI** [comment]
sadly i cannot let idea autoformat because then it would basically touch every file.. i think someone should do that at some point though.. and maybe get a official fabric codestyle together
i suggest the intellij idea defaults with star imports banned

**Comment by jamierocks** [comment]
There is one - see [fabric-docs](https://github.com/FabricMC/fabric-docs/blob/master/guidelines/intellij/FabricFormat.xml).

**Comment by NikkyAI** [comment]
there is 2 other options we have for implementing this

:one:. do dumb string replacements
`$VERSION` -> `0.1.0-12`

:two:. pass a `Map<String, Any>`
  and handle Strings, Numbers and Maps recursively
  this options would probably be unable to cleanly extend a already existing map in the json

**Comment by NikkyAI** [comment]
i implemented option :one: for now

**Comment by ChloeDawn** [comment]
Why is this needed for the json? Gradle provides `processResources` for stuff like this

**Comment by ChloeDawn** [comment]
```groovy
processResources {
  from(sourceSets.main.resources.srcDirs) {
    include 'mod.json'
    expand version: project.version
  }
  from(sourceSets.main.resources.srcDirs) {
    exclude 'mod.json'
  }
}
```
```json
{
  "id": "foobar",
  "name": "FooBar",
  "version": "$version",
  "side": "universal"
}
```

**Comment by NikkyAI** [comment]
This gets handled just fine by gradle iself just as @InsomniaKitten pointed out

so i see no reason to keep this PR open

**Pull request closed without merging**